File: 20190917/20190928/1002. 查找常用字符.cc
/*
 * Description: 查找常用字符
 * Author: AskeyNil
 * CreateDate: 2019-09-28 20:06:02
 * LastEditors: AskeyNil
 *
 * *********************************
 * **                             **
 * **     　  你只有足够努力     　  **
 * **       才能看上去毫不费力       **
 * **                             **
 * *********************************
 *
 */

#include <iostream>
#include <map>
#include <string>
#include <vector>

using namespace std;

class Solution {
  public:
    vector<string> commonChars(vector<string> &A) {
        vector<map<string, int>> counts;
        for (string &str : A) {
            map<string, int> m;
            for (char &c : str) {
                string s{c};
                m[s] += 1;
            }
            counts.push_back(m);
        }
        vector<string> result;
        // 遍历第一个字典
        for (auto &key : counts[0]) {
            int number = key.second;
            bool is_add = true;
            for (int i = 1; i < counts.size(); i++) {
                auto &value = counts[i];
                if (value.find(key.first) == value.end()) {
                    is_add = false;
                    break;
                } else {
                    if (value[key.first] < number) {
                        number = value[key.first];
                    }
                }
            }
            if (is_add) {
                for (int j = 0; j < number; j++) {
                    result.push_back(key.first);
                }
            }
        }
        return result;
    }
};

File: 20190917/20190928/1002. 查找常用字符.py
'''
Description: 查找常用字符
Author: AskeyNil
CreateDate: 2019-09-28 20:06:09
LastEditors: AskeyNil

*********************************
**                             **
**     　  你只有足够努力     　  **
**       才能看上去毫不费力       **
**                             **
*********************************

'''


class Solution:
    def commonChars(self, A: [str]) -> [str]:
        # 统计法
        counts = []
        for string in A:
            counts.append({})
            for char in string:
                if char in counts[-1]:
                    counts[-1][char] += 1
                else:
                    counts[-1][char] = 1
        result = []
        # 统计完了 遍历第一个字典
        for key in counts[0]:
            number = counts[0][key]
            is_add = True
            for dic in counts[0:]:
                if key not in dic:
                    is_add = False
                    break
                else:
                    if dic[key] < number:
                        number = dic[key]
            if is_add:
                for _ in range(number):
                    result.append(key)
        return result
